ancv/web/server.py

```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
+
+        log.debug("Starting client session.")
+        session = ClientSession()
+        app["client_session"] = session
+        log.debug("Started client session.")
+
+        log.debug("App context initialization done, yielding.")
+        yield
+
+        log.debug("App context teardown starting.")
+        await session.close()
+        log.debug("App context teardown done.")
+
+    async def fetch_and_render(self, session: ClientSession) -> str:
+        """Fetches the resume from URL and renders it.
+
+        Args:
+            session: The aiohttp client session to use for requests.
+
+        Returns:
+            The rendered template as string.
+        """
+        log = get_logger()
+        stopwatch = Stopwatch()
+        stopwatch(segment="Fetch Resume")
+
+        try:
+            async with session.get(self.destination) as response:
+                if response.status != 200:
+                    raise web.HTTPBadRequest(
+                        reason=f"Failed to fetch resume from {self.destination}"
+                    )
+                content = await response.text()
+                resume_json = json.loads(content)
+
+            with tempfile.NamedTemporaryFile(mode="w", suffix=".json") as tmp:
+                json.dump(resume_json, tmp)
+                tmp.flush()
+                template = Template.from_file(Path(tmp.name))
+                rendered = template.render()
+
+            log.info("Successfully refreshed resume data")
+            self._last_valid_render = rendered
+            return rendered
+
+        except Exception as e:
+            if hasattr(self, "_last_valid_render"):
+                log.warning(
+                    f"Failed to refresh resume, using cached version. Error: {str(e)}"
+                )
+                return self._last_valid_render
+            raise
+
+    async def root(self, request: web.Request) -> web.Response:
+        """The root endpoint, returning the rendered template with periodic refresh.
+
+        Args:
+            request: The incoming web request.
+
+        Returns:
+            The response containing the rendered resume.
+        """
+        log = LOGGER.bind(request=request)
+        session: ClientSession = request.app["client_session"]
+
+        current_time = time.time()
+        should_refresh = (
+            self.cache is None
+            or self.last_fetch is None
+            or (current_time - self.last_fetch) > self.refresh_interval
+        )
+
+        if should_refresh:
+            log.debug("Fetching fresh resume data.")
+            try:
+                self.cache = await self.fetch_and_render(session)
+                self.last_fetch = current_time
+            except Exception as e:
+                log.error("Failed to fetch resume", error=str(e))
+                if self.cache is None:
+                    raise
+
+        log.debug("Serving rendered template.")
+        return web.Response(text=self.cache)
```

Comment:
I think this won't be able to do type narrowing properly? I suspect `self.cache` is detected as `Optional[str]` here (do you have an editor which lets you check this?)?

If yes, let's move this into the `else` block of `if should_refresh`. That could/should allow `mypy` etc. to detect that `self.cache` is `str`, as `is None` was excluded at that point.

Reply:
Changed the logic when doing the function split, so now it's a `str`, not `Optional` anymore. =]